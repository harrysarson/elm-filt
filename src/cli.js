import meow from 'meow';
import log from 'log';
import logNode from 'log-node';

import fs from 'fs-extra';
import {detectElmVersion, filters, ElmFiltError} from './lib';

async function main() {
	logNode();

	const cli = meow(
		`
    Usage
      $ elm-filt <source> --keep <elm function>
    Examples
      $ elm-filt elm.js --keep Main.main --keep Main.update
      TODO
    `,
		{
			flags: {
				keep: {
					type: 'string'
				}
			}
		}
	);

	if (cli.input.length !== 1) {
		throw new ElmFiltError(`${
			cli.input.length
		} paths to JavaScript files generated by elm have be provided.
    Exactly 1 path must be provided. ${
			cli.input === []
				? ''
				: `Provided paths: ${['', ...cli.input].join('\n  - ')}`
		}`);
	}

	if (cli.flags.keep === undefined) {
		throw new ElmFiltError(
			'One or more fully qualified elm function names must be provided.'
		);
	}

	const keep = (x => (typeof x === 'string' ? [x] : x))(cli.flags.keep);
	const input = cli.input[0];

	const source = await (async () => {
		try {
			return await fs.readFile(input, 'utf8');
		} catch (error) {
			throw new ElmFiltError(
				`The path ${input} could not be read! Details:\n${error}`
			);
		}
	})();

	const version = (() => {
		try {
			return detectElmVersion(source);
		} catch (error) {
			if (error.supportedElmVersions === undefined) {
				throw new ElmFiltError(
					`Unkown error when attempting to recognise ${input}! Details:\n${error}`
				);
			} else {
				throw new ElmFiltError(`The JavaScript file at ${input} does not contain the output of a supported elm compiler.
The following elm compiler versions are supported:${[
					'',
					...error.supportedElmVersions
				].join('\n  - ')}`);
			}
		}
	})();

	const filtered = filters[version]({
		source,
		packageName: 'author/project',
		keeps: keep
	});

	for (const {elmIdentifier, javascript} of filtered) {
		process.stdout.write(`
// For ${elmIdentifier}
${javascript}
`);
	}
}

main().catch(error => {
	if (error.elmFiltError === true) {
		log.error(error.message);
		process.exit(2); // eslint-disable-line unicorn/no-process-exit
	} else {
		log.error(error);
		process.exit(3); // eslint-disable-line unicorn/no-process-exit
	}
});
