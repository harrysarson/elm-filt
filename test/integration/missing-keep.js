import test from 'ava';
import {exec} from '../helpers/cli';

test(exec, 'examples/elm-0.19.0', async (program, t) => {
	const output = await t.throwsAsync(program);
	t.is(output.code, 2);
	return {snapshot: output};
});
