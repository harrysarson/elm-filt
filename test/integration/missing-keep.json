{
    "args": "./examples/elm-0.19.0",
    "code": 2
}
