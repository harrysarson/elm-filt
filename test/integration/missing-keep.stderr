Ã— One or more fully qualified elm function names must be provided.
